<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>SeqFindr package &mdash; SeqFindr 0.34.0 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.34.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="SeqFindr 0.34.0 documentation" href="index.html" />
    <link rel="up" title="SeqFindr" href="modules.html" />
    <link rel="next" title="SeqFindr Developer HOWTO" href="developer.html" />
    <link rel="prev" title="SeqFindr" href="modules.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="developer.html" title="SeqFindr Developer HOWTO"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="modules.html" title="SeqFindr"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">SeqFindr 0.34.0 documentation</a> &raquo;</li>
          <li><a href="modules.html" accesskey="U">SeqFindr</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="seqfindr-package">
<h1>SeqFindr package<a class="headerlink" href="#seqfindr-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-SeqFindr.blast">
<span id="seqfindr-blast-module"></span><h2>SeqFindr.blast module<a class="headerlink" href="#module-SeqFindr.blast" title="Permalink to this headline">¶</a></h2>
<p>SeqFindr BLAST methods</p>
<dl class="function">
<dt id="SeqFindr.blast.make_BLAST_database">
<tt class="descclassname">SeqFindr.blast.</tt><tt class="descname">make_BLAST_database</tt><big>(</big><em>fasta_file</em><big>)</big><a class="headerlink" href="#SeqFindr.blast.make_BLAST_database" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a fasta_file, generate a nucleotide BLAST database</p>
<p>Database will end up in DB/ of working directory or OUTPUT/DB if an
output directory is given in the arguments</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fasta_file</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; full path to a fasta file</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">the strain id <strong>(must be delimited by &#8216;_&#8217;)</strong></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.blast.parse_BLAST">
<tt class="descclassname">SeqFindr.blast.</tt><tt class="descname">parse_BLAST</tt><big>(</big><em>blast_results</em>, <em>tol</em>, <em>careful</em><big>)</big><a class="headerlink" href="#SeqFindr.blast.parse_BLAST" title="Permalink to this definition">¶</a></dt>
<dd><p>Using NCBIXML parse the BLAST results, storing &amp; returning good hits</p>
<dl class="docutils">
<dt>Here good hits are:</dt>
<dd><ul class="first last simple">
<li>hsp.identities/float(record.query_length) &gt;= tol</li>
</ul>
</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>blast_results</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; full path to a blast run output file (in XML format)</li>
<li><strong>tol</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; the cutoff threshold (see above for explaination)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list of satifying hit names</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.blast.run_BLAST">
<tt class="descclassname">SeqFindr.blast.</tt><tt class="descname">run_BLAST</tt><big>(</big><em>query</em>, <em>database</em>, <em>args</em>, <em>cons_run</em><big>)</big><a class="headerlink" href="#SeqFindr.blast.run_BLAST" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a mfa of query sequences of interest &amp; a database, search for them.</p>
<dl class="docutils">
<dt>Important to note:</dt>
<dd><ul class="first last simple">
<li>Turns dust filter off,</li>
<li>Only a single target sequence (top hit),</li>
<li>Output in XML format as blast.xml.</li>
</ul>
</dd>
</dl>
<p># TODO: Add  evalue filtering ?
# TODO: add task=&#8217;blastn&#8217; to use blastn scoring ?</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">default is megablast</p>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">tblastx funcationality has not been checked</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>query</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; the fullpath to the vf.mfa</li>
<li><strong>database</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; the full path of the databse to search for the vf in</li>
<li><strong>args</strong> (<em>argparse args (dictionary)</em>) &#8211; the arguments parsed to argparse</li>
<li><strong>cons_run</strong> (<em>boolean</em>) &#8211; part of a mapping consensus run</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">the path of the blast.xml file</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-SeqFindr.config">
<span id="seqfindr-config-module"></span><h2>SeqFindr.config module<a class="headerlink" href="#module-SeqFindr.config" title="Permalink to this headline">¶</a></h2>
<p>SeqFindr configuration class: 100% test coverage, &gt; 9 PyLint score</p>
<dl class="class">
<dt id="SeqFindr.config.SeqFindrConfig">
<em class="property">class </em><tt class="descclassname">SeqFindr.config.</tt><tt class="descname">SeqFindrConfig</tt><big>(</big><em>alt_location=None</em><big>)</big><a class="headerlink" href="#SeqFindr.config.SeqFindrConfig" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></a></p>
<p>A SeqFindr configuration class - subtle manipulation to plots</p>
<dl class="method">
<dt id="SeqFindr.config.SeqFindrConfig.dump_items">
<tt class="descname">dump_items</tt><big>(</big><big>)</big><a class="headerlink" href="#SeqFindr.config.SeqFindrConfig.dump_items" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints all set configuration options to STDOUT</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="SeqFindr.config.read_config">
<tt class="descclassname">SeqFindr.config.</tt><tt class="descname">read_config</tt><big>(</big><em>alt_location</em><big>)</big><a class="headerlink" href="#SeqFindr.config.read_config" title="Permalink to this definition">¶</a></dt>
<dd><p>Read a SeqFindr configuration file</p>
<p>Currently only supports category colors in RGB format</p>
<p>category_colors = [(0,0,0),(255,255,255),....,(r,g,b)]</p>
</dd></dl>

</div>
<div class="section" id="module-SeqFindr.imaging">
<span id="seqfindr-imaging-module"></span><h2>SeqFindr.imaging module<a class="headerlink" href="#module-SeqFindr.imaging" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="SeqFindr.imaging.generate_colors">
<tt class="descclassname">SeqFindr.imaging.</tt><tt class="descname">generate_colors</tt><big>(</big><em>number_required</em>, <em>seed</em><big>)</big><a class="headerlink" href="#SeqFindr.imaging.generate_colors" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a list of length number of distinct &#8220;good&#8221; random colors</p>
<p>See: <a class="reference external" href="https://github.com/fmder/ghalton">https://github.com/fmder/ghalton</a></p>
<dl class="docutils">
<dt>Based on <a class="reference external" href="http://martin.ankerl.com/2009/12/09/">http://martin.ankerl.com/2009/12/09/</a></dt>
<dd>how-to-create-random-colors-programmatically/</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>number_required</strong> &#8211; int</li>
<li><strong>seed</strong> &#8211; the random seed</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Type:</th><td class="field-body"><p class="first">int</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Type:</th><td class="field-body"><p class="first">int</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a list of lists in the form: [[243, 137, 121], [232, 121, 243],
[216, 121, 243]]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.imaging.hsv_to_rgb">
<tt class="descclassname">SeqFindr.imaging.</tt><tt class="descname">hsv_to_rgb</tt><big>(</big><em>h</em>, <em>s</em>, <em>v</em><big>)</big><a class="headerlink" href="#SeqFindr.imaging.hsv_to_rgb" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert HSV to RGB</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>h</strong> &#8211; hue</li>
<li><strong>s</strong> &#8211; saturation</li>
<li><strong>v</strong> &#8211; value</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-SeqFindr.seqfindr">
<span id="seqfindr-seqfindr-module"></span><h2>SeqFindr.seqfindr module<a class="headerlink" href="#module-SeqFindr.seqfindr" title="Permalink to this headline">¶</a></h2>
<p>SeqFindr v0.34.0 - A tool to easily create informative genomic feature plots (<a class="reference external" href="http://github.com/mscook/SeqFindr">http://github.com/mscook/SeqFindr</a>)</p>
<dl class="function">
<dt id="SeqFindr.seqfindr.build_matrix_row">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">build_matrix_row</tt><big>(</big><em>all_vfs</em>, <em>accepted_hits</em>, <em>score=None</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.build_matrix_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Populate row given all possible hits, accepted hits and an optional score</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>all_vfs</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; a list of all virulence factor ids</li>
<li><strong>accepted_hits</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; a list of a hits that passed the cutoof</li>
<li><strong>score</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; the value to fill the matrix with (default = None which
implies 0.5)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a list of floats</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.check_singularity">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">check_singularity</tt><big>(</big><em>matrix</em>, <em>cons</em>, <em>invert</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.check_singularity" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if there are any informative sites in the matrix</p>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.cluster_matrix">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">cluster_matrix</tt><big>(</big><em>matrix</em>, <em>labels</em>, <em>dpi</em>, <em>by_cols</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.cluster_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>From a matrix, generate a distance matrix &amp; perform hierarchical clustering</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>matrix</strong> (<em>numpy matrix</em>) &#8211; a numpy matrix of scores</li>
<li><strong>labels</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; the ids for all row elements or column elements</li>
<li><strong>dpi</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; the resolution to save the diagram at</li>
<li><strong>by_cols</strong> (<em>boolean (default == False)</em>) &#8211; whether to perform the clustering by row similarity
(default) or column similarity.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tuple of the updated (clustered) matrix &amp; the updated labels</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.core">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">core</tt><big>(</big><em>args</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.core" title="Permalink to this definition">¶</a></dt>
<dd><p>The &#8216;core&#8217; SeqFindr method</p>
<p>TODO: Exception handling if do_run fails or produces no results</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> &#8211; the arguments given from argparse</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.determine_nohit_score">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">determine_nohit_score</tt><big>(</big><em>cons</em>, <em>invert</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.determine_nohit_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine the value in the matrix assigned to nohit given SeqFindr options</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>cons</strong> (<em>boolean</em>) &#8211; whether the Seqfindr run is using mapping consensus data
or not</li>
<li><strong>invert</strong> &#8211; whether the Seqfindr run is inverting (missing hits to
be shown as black bars.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">the value defined as no hit in the results matrix</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.do_run">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">do_run</tt><big>(</big><em>args</em>, <em>data_path</em>, <em>match_score</em>, <em>vfs_list</em>, <em>cons_run</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.do_run" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a SeqFindr run</p>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.match_matrix_rows">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">match_matrix_rows</tt><big>(</big><em>ass_mat</em>, <em>cons_mat</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.match_matrix_rows" title="Permalink to this definition">¶</a></dt>
<dd><p>Reorder a second matrix based on the first row element of the 1st matrix</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>ass_mat</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; a 2D list of scores</li>
<li><strong>cons_mat</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; a 2D list scores</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">2 matricies (2D lists)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.plot_matrix">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">plot_matrix</tt><big>(</big><em>matrix</em>, <em>strain_labels</em>, <em>vfs_classes</em>, <em>gene_labels</em>, <em>show_gene_labels</em>, <em>color_index</em>, <em>config_object</em>, <em>grid</em>, <em>seed</em>, <em>dpi</em>, <em>size</em>, <em>svg</em>, <em>cluster_column</em>, <em>aspect='auto'</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.plot_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the VF hit matrix</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>matrix</strong> &#8211; the numpy matrix of scores</li>
<li><strong>strain_labels</strong> &#8211; the strain (y labels)</li>
<li><strong>vfs_classes</strong> &#8211; the VFS class (in mfa header [class])</li>
<li><strong>gene_labels</strong> &#8211; the gene labels</li>
<li><strong>show_gene_labels</strong> &#8211; wheter top plot the gene labels</li>
<li><strong>color_index</strong> &#8211; for a single class, choose a specific color</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.prepare_queries">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">prepare_queries</tt><big>(</big><em>args</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.prepare_queries" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a set of sequences of interest, extract all query &amp; query classes</p>
<p>A sequence of interest file is a mfa file in the format:</p>
<p>&gt;ident, gene id, annotation, organism [class]</p>
<p>query = gene id
query_class = class</p>
<p>Location of sequence of interest file is defined by args.seqs_of_interest</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<em>argparse args</em>) &#8211; the argparse args containing args.seqs_of_interest
(fullpath) to a sequence of interest DB (mfa file)</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">2 lists, 1) of all queries and, 2) corresponding query classes</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.strip_bases">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">strip_bases</tt><big>(</big><em>args</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.strip_bases" title="Permalink to this definition">¶</a></dt>
<dd><p>Strip the 1st and last &#8216;N&#8217; bases from mapping consensuses</p>
<dl class="docutils">
<dt>Uses:</dt>
<dd><ul class="first last simple">
<li>args.cons</li>
<li>args.seqs_of_interest</li>
<li>arg.strip</li>
</ul>
</dd>
</dl>
<p>To avoid the effects of lead in and lead out coverage resulting in
uncalled bases</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<em>argparse args</em>) &#8211; the argparse args containing args.strip value</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">the updated args to reflect the args.cons &amp;
args.seqs_of_interest location</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.strip_id_from_matrix">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">strip_id_from_matrix</tt><big>(</big><em>mat</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.strip_id_from_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove the ID (1st row element) form a matrix</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>mat</strong> &#8211; a 2D list</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">a 2D list with the 1st row elelemnt (ID) removed</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.seqfindr.strip_uninteresting">
<tt class="descclassname">SeqFindr.seqfindr.</tt><tt class="descname">strip_uninteresting</tt><big>(</big><em>matrix</em>, <em>query_classes</em>, <em>query_list</em>, <em>cons</em>, <em>invert</em><big>)</big><a class="headerlink" href="#SeqFindr.seqfindr.strip_uninteresting" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove any columns where all elements in every position are absent</p>
<p>Also handles the query classes and x_lables.</p>
<div class="admonition attention">
<p class="first admonition-title">Attention</p>
<p class="last">new feature added in version 0.4.0</p>
</div>
<p>Toogle using: <strong>args.remove_empty_cols</strong></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>matrix</strong> &#8211; the SeqFindr hit matrix</li>
<li><strong>query_classes</strong> &#8211; a list of query classes</li>
<li><strong>query_list</strong> &#8211; a query list (x labels)</li>
<li><strong>cons</strong> &#8211; whether the Seqfindr run is using mapping consensus data
or not</li>
<li><strong>invert</strong> &#8211; whether the Seqfindr run is inverting (missing hits to
be shown as black bars.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tuple with three elements which are the: updated SeqFindr
matrix, the updated query_classes list and the updated
query_list respectively.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-SeqFindr.util">
<span id="seqfindr-util-module"></span><h2>SeqFindr.util module<a class="headerlink" href="#module-SeqFindr.util" title="Permalink to this headline">¶</a></h2>
<p>SeqFindr utility methods</p>
<dl class="function">
<dt id="SeqFindr.util.check_database">
<tt class="descclassname">SeqFindr.util.</tt><tt class="descname">check_database</tt><big>(</big><em>database_file</em><big>)</big><a class="headerlink" href="#SeqFindr.util.check_database" title="Permalink to this definition">¶</a></dt>
<dd><p>Check the database conforms to the SeqFindr format</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">this is not particulalry extensive</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name" colspan="2">Args database_file:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">full path to a database file as a string</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.util.del_from_list">
<tt class="descclassname">SeqFindr.util.</tt><tt class="descname">del_from_list</tt><big>(</big><em>target</em>, <em>index_positions</em><big>)</big><a class="headerlink" href="#SeqFindr.util.del_from_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes the elements in a list given by a index_positions list</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>target</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; a target list to have items removed</li>
<li><strong>index_positions</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; a list of index positions to be removed from
the target list</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a list with the elements removed defined by the index_positions
list</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.util.ensure_paths_for_args">
<tt class="descclassname">SeqFindr.util.</tt><tt class="descname">ensure_paths_for_args</tt><big>(</big><em>args</em><big>)</big><a class="headerlink" href="#SeqFindr.util.ensure_paths_for_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Ensure all arguments with paths are absolute &amp; have simplification removed</p>
<p>Just apply os.path.abspath &amp; os.path.expanduser</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> &#8211; the arguments given from argparse</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">an updated args</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.util.get_fasta_files">
<tt class="descclassname">SeqFindr.util.</tt><tt class="descname">get_fasta_files</tt><big>(</big><em>data_path</em><big>)</big><a class="headerlink" href="#SeqFindr.util.get_fasta_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all files ending with .fas/.fa/fna in a directory</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>data_path</strong> &#8211; the full path to the directory of interest</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a list of fasta files (valid extensions: .fas, .fna, .fa</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.util.init_output_dirs">
<tt class="descclassname">SeqFindr.util.</tt><tt class="descname">init_output_dirs</tt><big>(</big><em>output_dir</em><big>)</big><a class="headerlink" href="#SeqFindr.util.init_output_dirs" title="Permalink to this definition">¶</a></dt>
<dd><p>Create the output base (if needed) and change dir to it</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> &#8211; the arguments given from argparse</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.util.is_protein">
<tt class="descclassname">SeqFindr.util.</tt><tt class="descname">is_protein</tt><big>(</big><em>fasta_file</em><big>)</big><a class="headerlink" href="#SeqFindr.util.is_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if a FASTA file is protein or nucleotide.</p>
<p>Will return -1 if no protein detected</p>
<p>TODO: Abiguity characters?
TODO: exception if mix of protein/nucleotide?</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fasta_file</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; path to input FASTA file</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">number of protein sequences in fasta_file (int)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="SeqFindr.util.order_inputs">
<tt class="descclassname">SeqFindr.util.</tt><tt class="descname">order_inputs</tt><big>(</big><em>order_index_file</em>, <em>dir_listing</em><big>)</big><a class="headerlink" href="#SeqFindr.util.order_inputs" title="Permalink to this definition">¶</a></dt>
<dd><p>Given an order index file, maintain this order in the matrix plot</p>
<p><strong>This implies no clustering.</strong> Typically used when you already have
a phlogenetic tree.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>order_index_file</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; full path to a ordered file (1 entry per line)</li>
<li><strong>dir_listing</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; a listing from util.get_fasta_files</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list of updated glob.glob dir listing to match order specified</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-SeqFindr.vfdb_to_seqfindr">
<span id="seqfindr-vfdb-to-seqfindr-module"></span><h2>SeqFindr.vfdb_to_seqfindr module<a class="headerlink" href="#module-SeqFindr.vfdb_to_seqfindr" title="Permalink to this headline">¶</a></h2>
<div class="section" id="vfdb-to-seqfindr">
<h3>vfdb_to_seqfindr<a class="headerlink" href="#vfdb-to-seqfindr" title="Permalink to this headline">¶</a></h3>
<p>Convert VFDB formatted files (or like) to SeqFindr formatted database files</p>
<p>VFDB: Virulence Factors Database
www.mgc.ac.cn/VFs/
a reference database for bacterial virulence factors.</p>
<p>This is based on a sample file (TOTAL_Strep_VFs.fas) provided by
Nouri Ben Zakour.</p>
<p>Examples:</p>
<div class="highlight-python"><div class="highlight"><pre># Default (will set VFDB classification identifiers as the classification)
$ vfdb_to_seqfindr -i TOTAL_Strep_VFs.fas -o TOTAL_Strep_VFs.sqf

# Sets any classification to blank ([ ])
$ vfdb_to_seqfindr -i TOTAL_Strep_VFs.fas -o TOTAL_Strep_VFs.sqf -b

# Reads a user defined classification. 1 per in same order as input
# sequences
$ python convert_vfdb_to_SeqFindr.py -i TOTAL_Strep_VFs.fas
  -o TOTAL_Strep_VFs.sqf -c blah.dat
</pre></div>
</div>
<div class="section" id="about-option-class-file">
<h4>About option &#8211;class_file<a class="headerlink" href="#about-option-class-file" title="Permalink to this headline">¶</a></h4>
<p>Suppose you want to annotate a VF class with user defined values. Simply
develop a file containing the scheme (1-1 matching). If you had 6 input
sequences and the first 3 are Fe transporters and the next two are Toxins
and the final sequence is Misc your class file would look like this:</p>
<p>Fe transporter
Fe transporter
Fe transporter
Toxins
Toxins
Misc</p>
<dl class="function">
<dt id="SeqFindr.vfdb_to_seqfindr.main">
<tt class="descclassname">SeqFindr.vfdb_to_seqfindr.</tt><tt class="descname">main</tt><big>(</big><em>args</em><big>)</big><a class="headerlink" href="#SeqFindr.vfdb_to_seqfindr.main" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="SeqFindr.vfdb_to_seqfindr.order_by_class">
<tt class="descclassname">SeqFindr.vfdb_to_seqfindr.</tt><tt class="descname">order_by_class</tt><big>(</big><em>args</em><big>)</big><a class="headerlink" href="#SeqFindr.vfdb_to_seqfindr.order_by_class" title="Permalink to this definition">¶</a></dt>
<dd><p>Ensure that all particualr classes are in the same block</p>
</dd></dl>

</div>
</div>
</div>
<div class="section" id="module-SeqFindr">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-SeqFindr" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">SeqFindr package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-SeqFindr.blast">SeqFindr.blast module</a></li>
<li><a class="reference internal" href="#module-SeqFindr.config">SeqFindr.config module</a></li>
<li><a class="reference internal" href="#module-SeqFindr.imaging">SeqFindr.imaging module</a></li>
<li><a class="reference internal" href="#module-SeqFindr.seqfindr">SeqFindr.seqfindr module</a></li>
<li><a class="reference internal" href="#module-SeqFindr.util">SeqFindr.util module</a></li>
<li><a class="reference internal" href="#module-SeqFindr.vfdb_to_seqfindr">SeqFindr.vfdb_to_seqfindr module</a><ul>
<li><a class="reference internal" href="#vfdb-to-seqfindr">vfdb_to_seqfindr</a><ul>
<li><a class="reference internal" href="#about-option-class-file">About option &#8211;class_file</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-SeqFindr">Module contents</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="modules.html"
                        title="previous chapter">SeqFindr</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="developer.html"
                        title="next chapter">SeqFindr Developer HOWTO</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/SeqFindr.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="developer.html" title="SeqFindr Developer HOWTO"
             >next</a> |</li>
        <li class="right" >
          <a href="modules.html" title="SeqFindr"
             >previous</a> |</li>
        <li><a href="index.html">SeqFindr 0.34.0 documentation</a> &raquo;</li>
          <li><a href="modules.html" >SeqFindr</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013-2014, Mitchell Stanton-Cook, Nabil Alikhan &amp; Hamza Khan.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
  </body>
</html>